File: cpp/utils/control.cpp
#include "control.h"

template <typename T>
void PIDController<T>::Init() {
    // last_time = std::chrono::time_point_cast<std::chrono::microseconds>(std::chrono::system_clock::now()).time_since_epoch().count();
    integral_error = T::Zero();
    last_error = T::Zero();
}

template <typename T>
T PIDController<T>::Step(float dt, T error) {
    T P = kp * error;

    integral_error += error * dt;
    T I = ki * integral_error;

    T D = kd * (error - last_error) / dt;
    last_error = error;

    return P + I + D;
}

template <typename T>
T PIDController<T>::Step(float dt, T error, T error_dot) {
    T P = kp * error;

    integral_error += error * dt;
    T I = ki * integral_error;

    T D = kd * error_dot;

    return P + I + D;
}

template class PIDController<Vector3f>;
template class PIDController<Vector2f>;

File: py/control.py
import numpy as np
import time


class PIDController:
    def __init__(self, Kp, Ki, Kd, dim):
        self.Kp = Kp
        self.Ki = Ki
        self.Kd = Kd
        self.dim = dim
    
    
    def start(self):
        self.E_acc = np.zeros(self.dim)
        self.last_time = time.time()
        self.last_error = np.zeros(self.dim)
    
    
    def step(self, error, error_dot):
        dt = time.time() - self.last_time
        self.last_time = time.time()
        
        P = self.Kp * error
        
        self.E_acc += error * dt
        I = self.Ki * self.E_acc
        
        if error_dot is None:
            derror = (error - self.last_error)
            self.last_error = error
            D = self.Kd * derror / dt
        else:
            D = self.Kd * error_dot
        
        return P + I + D

